File: validation/generate_validation_data.py
"""
This scripts uses graph_tool to generate the data which will be compared to
  the ones obtained with edgelist_perco_t, namely
  - get_nb_components
  - get_size_largest_perco_component
  - get_size_second_largest_perco_component

Copyright (C) 2020  Antoine Allard (antoineallard.info)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
import graph_tool.all as gt
import numpy as np

# Number of vertices
nb_vertices = 250

# Expected average degree
expected_average_degree = 5

# Sampler for the degree of the vertices
def sample_deg():
    return np.random.poisson(expected_average_degree)


# Generates a graph
g = gt.random_graph(nb_vertices, sample_deg, directed=False)

# Saves the graph as an edgelist.
with open("validation_edgelist.dat", "ab") as f:
    np.savetxt(f, g.get_edges(),
               fmt="%15s", delimiter=" ",
               header=" SourceVertex    TargetVertex")

with open("validation_raw_results.dat", "w" ) as f:

    # g.set_fast_edge_removal(True)
    f.write('#   nb_vertices        nb_edges        size_1st        size_2nd         nb_comp\n')

    for i in range(25):

        g2 = g.copy()
        g2.set_fast_edge_removal(fast=True)
        edges = g2.get_edges()
        nb_edges = len(edges)
        np.random.shuffle(edges)

        ncomp = np.zeros(nb_edges + 1)
        size1 = np.zeros(nb_edges + 1)
        size2 = np.zeros(nb_edges + 1)

        comps = sorted(gt.label_components(g2)[1], reverse=True)
        ncomp[0] = len(comps)
        size1[0] = comps[0]
        if ncomp[0] > 1:
             size2[0] = comps[1]

        for e in np.arange(0, nb_edges, 1):
            g2.remove_edge(g2.edge(edges[e][0], edges[e][1]))
            comps = sorted(gt.label_components(g2)[1], reverse=True)
            ncomp[e+1] = len(comps)
            size1[e+1] = comps[0]
            if ncomp[e+1] > 1:
                 size2[e+1] = comps[1]

        np.savetxt(f, np.column_stack((np.ones((nb_edges + 1, 1)) * nb_vertices, range(nb_edges, -1, -1), size1, size2, ncomp)),
                   delimiter=' ', encoding='utf-8',
                   fmt='%15d %15d %15d %15d %15d ')

File: validation/generate_validation_data.cpp
/*
 *
 *  This code generates the text files used to validate the edgelist_perco_t
 *    class of the AGL to analyze the structure of undirected, unweighted and simple graph.
 *
 *  Compilation requires the c++11 standard.
 *    Example: g++ -O3 -std=c++11 generate_validation_data.cpp -o generate_validation_data
 *
 *  Author:  Antoine Allard
 *  WWW:     antoineallard.info
 *  Date:    November 2020
 *
 *
 *  Copyright (C) 2020 Antoine Allard
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */


 // Standard template library
 #include <iomanip>
 #include <fstream>
 #include <string>
 // Portable graph library
 #include "../src/edgelist_perco_t.hpp"


 int main(int argc, char const *argv[])
 {
  // Opens the file stream.
  std::ofstream output_file("validation_edgelist_perco_t.dat");

  // column width in the output file.
  int width = 15;

  // Parameters of the simulations
  int nb_simulations = 25000;
  double dt = 0.01;

  // Writes the header.
  output_file << "#";
  output_file << std::setw(width - 1) << "edge_prob" << " ";
  output_file << std::setw(width) << "nb_vertices" << " ";
  output_file << std::setw(width) << "nb_edges" << " ";
  output_file << std::setw(width) << "size_1st" << " ";
  output_file << std::setw(width) << "size_2nd" << " ";
  output_file << std::setw(width) << "nb_comp" << " ";
  output_file << std::endl;

  // Loads the random edgelist.
  agl::edgelist_perco_t g("validation_edgelist.dat");

  // Performs several bond percolation simulations for each shuffled graph.
  for(int m(0), nb_edges; m<nb_simulations; ++m)
  {
    for(double t(0), tt(1.00001); t<=tt; t+=dt)
    {
      // Generates a bond percolated instance.
      nb_edges = g.bond_percolate(t);

      // Outputs the raw results.
      output_file << std::setw(width) << t << " ";
      output_file << std::setw(width) << g.get_nb_vertices() << " ";
      output_file << std::setw(width) << nb_edges << " ";
      output_file << std::setw(width) << g.get_size_largest_perco_component() << " ";
      output_file << std::setw(width) << g.get_size_second_largest_perco_component() << " ";
      output_file << std::setw(width) << g.get_nb_components() << " ";
      output_file << std::endl;
    }
  }

  // Closes the file stream.
  output_file.close();


  return 0;
 }
